### ISSUE TITLE
[html-aam] Computing a role for the “aside” element creates an infinite loop

### ISSUE THREAD

**Issue by sideshowbarker** [w3c/aria#2391]
Consider these two things:

**A. Computing a role for an `aside` element.** The conditions at https://w3c.github.io/html-aam/#el-aside for computing a role for the `aside` element in turn require computing an accessible name for the `aside` element.

**B. Computing an accessible name for an element.** Multiple steps in the algorithm at https://w3c.github.io/accname/#computation-steps for computing an accessible name for an element in turn require computing a role for that element.

That is: In order to implement **A. Computing a role for an aside element** in my code, I need to have that code call my implementation of **B. Computing an accessible name for an element.** But in order to implement **B. Computing an accessible name for an element** in my code, I need to have that code call my implementation of **A. Computing a role for an aside element**.

In other words, my code gets into an infinite loop if I implement the requirements as specified.

Link to documentation: https://w3c.github.io/html-aam/#el-aside

Does the issue exists in the editors draft is the most recent draft of the specification? Yes

**Comment by smockle** [comment]
> **B.**…computing an accessible name for an element in turn require computing a role for that element.

`aside` has either a `complementary` or `generic` role. Since `generic` [can’t be named](https://w3c.github.io/aria/#namefromprohibited), when checking whether `aside` has an accessible name, its role can be assumed to be `complementary`, right? By assuming that, actual role computation would be unnecessary, avoiding the infinite loop.

**Comment by scottaohara** [comment]
would this make more sense if the spec indicated the default roles to be complementary / region - and THEN have a condition to change those roles based on ancestral checks (for aside) or if the elements are named (for both elements) - rather than separating these out so the elements each had two tables?

**Comment by annevk** [comment]
I think you should resolve the loop in the algorithms somehow. E.g., give "computing an accessible name" a boolean that indicates it's invoked for role computing purposes in which case it uses fallback roles for certain elements might do the trick?
